dali/imgcodec/decoders/nvjpeg.cc

```diff
@@ -0,0 +1,169 @@
+  device_allocator_(nvjpeg_memory::GetDeviceAllocator()),
+  pinned_allocator_(nvjpeg_memory::GetPinnedAllocator()),
+  resources_(tp->NumThreads()) {
+  CUDA_CALL(nvjpegCreateSimple(&nvjpeg_handle_));
+
+  for (auto &resource : resources_) {
+    CUDA_CALL(nvjpegJpegStreamCreate(nvjpeg_handle_, &resource.jpeg_stream));
+    CUDA_CALL(nvjpegBufferPinnedCreate(nvjpeg_handle_, &pinned_allocator_,
+                                       &resource.pinned_buffer));
+    CUDA_CALL(nvjpegBufferDeviceCreate(nvjpeg_handle_, &device_allocator_,
+                                       &resource.device_buffer));
+    CUDA_CALL(cudaStreamCreateWithFlags(&resource.stream, cudaStreamNonBlocking));
+
+    CUDA_CALL(cudaEventCreate(&resource.decode_event));
+    CUDA_CALL(cudaEventRecord(resource.decode_event, resources_[0].stream));
+
+    auto &decoder_data = resource.decoder_data;
+    auto backend = NVJPEG_BACKEND_HYBRID;  // TODO(msala) allow other backens
+    CUDA_CALL(nvjpegDecoderCreate(nvjpeg_handle_, backend, &decoder_data.decoder));
+    CUDA_CALL(nvjpegDecoderStateCreate(nvjpeg_handle_, decoder_data.decoder, &decoder_data.state));
+  }
+}
+
+NvJpegDecoderInstance::~NvJpegDecoderInstance() {
+  for (auto &thread_id : tp_->GetThreadIds()) {
+    nvjpeg_memory::DeleteAllBuffers(thread_id);
+  }
+
+  for (const auto &resource : resources_) {
+    CUDA_CALL(cudaStreamSynchronize(resource.stream));
+
+    CUDA_CALL(nvjpegJpegStreamDestroy(resource.jpeg_stream));
+    CUDA_CALL(nvjpegBufferPinnedDestroy(resource.pinned_buffer));
+    CUDA_CALL(nvjpegBufferDeviceDestroy(resource.device_buffer));
+    CUDA_CALL(cudaEventDestroy(resource.decode_event));
+    CUDA_CALL(cudaStreamDestroy(resource.stream));
+
+    CUDA_CALL(nvjpegDecoderDestroy(resource.decoder_data.decoder));
+    CUDA_CALL(nvjpegJpegStateDestroy(resource.decoder_data.state));
+  }
+
+  CUDA_CALL(nvjpegDestroy(nvjpeg_handle_));
+}
+
+void NvJpegDecoderInstance::SetParam(const char *name, const any &value) {
+  if (strcmp(name, "device_memory_padding") == 0) {
+    device_memory_padding_ = any_cast<size_t>(value);
+
+    if (device_memory_padding_ > 0) {
+      for (auto thread_id : tp_->GetThreadIds()) {
+        nvjpeg_memory::AddBuffer<mm::memory_kind::device>(thread_id, device_memory_padding_);
+      }
+    }
+  } else if (strcmp(name, "host_memory_padding") == 0) {
+    host_memory_padding_ = any_cast<size_t>(value);
+
+    if (host_memory_padding_ > 0) {
+      for (auto thread_id : tp_->GetThreadIds()) {
+        nvjpeg_memory::AddHostBuffer(thread_id, host_memory_padding_);
+        nvjpeg_memory::AddHostBuffer(thread_id, host_memory_padding_);
+      }
+    }
+  }
+}
+
+any NvJpegDecoderInstance::GetParam(const char *name) const {
+  if (strcmp(name, "device_memory_padding") == 0) {
+    return device_memory_padding_;
+  } else if (strcmp(name, "host_memory_padding") == 0) {
+    return host_memory_padding_;
+  } else {
+    DALI_FAIL("Unrecognized param name");
```

Comment:
As already mentioned in slack, it should not fail.
```suggestion
    return {};
```